### PR TITLE
Taught FileDownloader to correctly handle on-demand remotes.

### LINKED ISSUE #4019: --url file:///foo/bar/PULP_MANIFEST with policy of on_demand or streamed raises unhandled RuntimeError
**Version**

Using the all-in-one pulp container with pulpcore 3.28.1 and pulp-file 1.14.3

**Describe the bug**

With a remote with --url file:///foo/bar/ using either on_demand or streamed policy, when attempting to retrieve a file via wget, a 500 Internal Server Error is returned, and the logs show an unhandled RunTimeError

**To Reproduce**

```
pulp file remote create --name foobar --url file:///foo/bar/PULP_MANIFEST --policy on_demand
...
wget http://myserver/pulp/content/foobar/somefile
...
HTTP request sent, awaiting response... 500 Internal Server Error
2023-07-08 01:19:34 ERROR 500: Internal Server Error.
```

**Expected behavior**

The file is successfully downloaded

**Additional context**

I mentioned this issue in IRC and it was confirmed as reproducible

### PR DESCRIPTION
**Pull request by ggainey**
fixes #4019 .

### CODE REVIEW

**Comment by ggainey** [comment]
Needs a test, getting it up to get some clean test runs and more eyes.

File: pulpcore/download/file.py [hunk 1]
```diff
@@ -37,6 +37,9 @@ def __init__(self, url, *args, **kwargs):
         RemoteSerializer().validate_url(url)
         p = urlparse(url)
         self._path = os.path.abspath(os.path.join(p.netloc, p.path))
+        if kwargs.get("headers_ready_callback", None):
+            self.headers_ready_callback = kwargs.get("headers_ready_callback")
```
Comment (mdellweg) [inline pulpcore/download/file.py:41]: It looks like it's perfectly safe to have `headers_ready_callback==None` (also `get` defaults to `None` anyway).
```suggestion
        self.headers_ready_callback = kwargs.get("headers_ready_callback")
```

**Comment by ggainey** [comment]
Status: writing a test discovered that there is more work to be done.

File: pulpcore/download/file.py (hunk 1 above)
Comment (ggainey) [reply pulpcore/download/file.py]: That code turned out to not be the actual issue (alas)

File: pulpcore/content/handler.py [hunk 2]
```diff
@@ -1012,6 +1012,8 @@ async def handle_response_headers(headers):
 
         async def handle_data(data):
             nonlocal data_size_handled
+            if not response.prepared:
+                await response.prepare(request)
```
Comment (mdellweg) [inline pulpcore/content/handler.py:1016]: The result of preparing a request is that you cannot add headers anymore (as far as I remember).
Is handle_response_headers not executed in the use case you are fixing here? Or is there a chance it didn't finish yet?
Can we use a `asyncio.Event` to make this function wait until it is set by `handle_response_headers` after the `prepare` call?

**Review by mdellweg — CHANGES_REQUESTED** [review]

File: pulpcore/content/handler.py (hunk 2 above)
Comment (ggainey) [reply pulpcore/content/handler.py:1016]: The path for returning an on-demand artifact, from a file: entity, through the content-app, is tortuous. handle_response_headers never gets invoked. If you've gotten to handle_data(), the content-app "assumes" prepare() has completed. If you are in this call, you're about to start writing, and will fail if prepare() hasn't happened.

This change makes it at least possible to serve on_demand file: artifacts. A better answer would be rearchitecting/rewriting the content-app to not be a mass of barbed wire - but fixing this use-case isn't the place for that.

Comment (mdellweg) [reply pulpcore/content/handler.py:1016]: OK, I understand now. This is the `FileDownloader` not even knowing about `headers_ready_callback` and therefore never bothering to call it. With the `HttpDownloader` on the other hand, we can always assume that `headers_ready_callback` is awaited before `handle_data` is ever invoked.
Please leave a description like this in a comment there.

File: pulpcore/tests/functional/api/using_plugin/test_content_access.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+"""Tests related to content delivery."""
+import pytest
+import uuid
+
+from pulp_smash.pulp3.utils import (
+    gen_distribution,
+)
```
Comment (mdellweg) [inline pulpcore/tests/functional/api/using_plugin/test_content_access.py:7]: Please don't use pulp-smash.
There's a `file_distribution_factory` for this.

**Review by mdellweg — CHANGES_REQUESTED** [review]

**Review by mdellweg — APPROVED** [review]

**Review by dkliban — APPROVED** [review]

**Pull request merged**